src/ServiceControl.Monitoring/Infrastructure/RemoveExpiredEndpointInstances.cs

```diff
@@ -0,0 +1,56 @@
+﻿namespace ServiceControl.Monitoring;
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Http.Diagrams;
+using Infrastructure;
+using Infrastructure.Api;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+
+public class RemoveExpiredEndpointInstances(
+    ILogger<RemoveExpiredEndpointInstances> logger,
+    IEndpointMetricsApi metricsApi,
+    EndpointInstanceActivityTracker activityTracker,
+    TimeProvider timeProvider) : BackgroundService
+{
+    const int IntervalInMinutes = 5;
+
+    protected override async Task ExecuteAsync(CancellationToken cancellationToken)
+    {
+        logger.LogInformation($"Starting {nameof(RemoveExpiredEndpointInstances)}");
+
+        try
+        {
+            using PeriodicTimer timer = new(TimeSpan.FromMinutes(IntervalInMinutes), timeProvider);
+
+            do
+            {
+                try
+                {
+                    foreach (MonitoredEndpoint monitoredEndpoint in metricsApi.GetAllEndpointsMetrics())
+                    {
+                        foreach (string endpointInstanceId in monitoredEndpoint.EndpointInstanceIds)
+                        {
+                            if (activityTracker.IsExpired(new EndpointInstanceId(monitoredEndpoint.Name,
+                                    endpointInstanceId)))
+                            {
+                                metricsApi.DeleteEndpointInstance(monitoredEndpoint.Name, endpointInstanceId);
```

Comment:
If the Endpoint Instance is removed, will this affect the 30 days' worth of historical data gathered by the licensing component?

Reply:
Good question @awright18 
It is not affected, we collect that data every 5 minutes, so whatever we get that is what we send to the error instance, the error instance saved the data in storage.
So when an endpoint instance is removed, it just means we no longer send data for that endpoint instance which is pretty much what we are currently doing except we currently send zero.